spec/Classes.tex

```diff
@@ -663,73 +604,456 @@ \subsection{User-Defined Constructors}
+\begin{chapeloutput}
+fieldsOutOfOrder.chpl:12: In initializer:
+fieldsOutOfOrder.chpl:17: error: Field "y" initialized out of order
+fieldsOutOfOrder.chpl:17: note: initialization of fields before .init() call must be in field declaration order
+\end{chapeloutput}
+
+The first initializer leverages the well-defined order of omitted field
+initialization to use the default value of field \chpl{y} in order to
+explicitly initialize field \chpl{z}.
+
+The second initializer initializes field \chpl{z} before field \chpl{y},
+causing a compile-time error to be issued.
+
+\end{chapelexample}
+
+\begin{rationale}
+Without this rule the compiler could insert default initialization for field
+\chpl{y} before \chpl{z} is explicitly initialized. The following statement
+would then be {\em assignment} to field \chpl{y}, despite appearing to be
+initialization. This subtle difference may be confusing and surprising, and
+is avoided by requiring fields to be initialized in field declaration order.
+\end{rationale}
+
+\subsubsection{The complete Method}
+\label{The_Complete_Method}
+\index{classes!initializers!user-defined!complete}
+
+As the initializer makes progress, the class instance is incrementally
+initialized. In order to prevent usage of uninitialized memory, there are
+restrictions on usage of the class instance before it is fully initialized:
+
+\begin{itemize}
+\item Methods may not be invoked on partially-initialized instances
+\item \chpl{this} may not be passed to functions while partially-initialized
+\end{itemize}
+
+These rules allow all methods and functions to assume that class instances
+have been initialized, provided their value is not \chpl{nil}.
+
+\begin{rationale}
+  The compiler could conceivably attempt to analyze methods and functions to
+  determine which fields are used, and selectively allow method calls on
+  partially-initialized class instances. A simpler rule, which is easier to
+  remember and implement, is to forbid method calls on partially-initialized
+  instances.
+\end{rationale}
+
+Methods may be called and \chpl{this} may be passed to functions only after the
+built-in \chpl{complete} method is invoked.
+
+\begin{rationale}
+Due to support for omitted field initialization, there is potential for
+confusion regarding the overall status of initialization. This confusion is
+addressed in the design by requiring \chpl{complete} to explicitly mark the
+transition between partially and fully initialized instances.
+\end{rationale}
+
+This method may not be overridden. If any fields have not been initialized by
+the time the \chpl{complete} method is invoked, they will be considered omitted
+and the compiler will insert initialization statements as described earlier. If
+the user does not invoke the \chpl{complete} method explicitly, the compiler
+will insert a call to \chpl{complete} at the end of the initializer. Note that
+even if the user explicitly initializes every field, the \chpl{complete} method
```

Comment:
I would split this into a separate paragraph.  We can have "note" sections, right?